### PR TITLE
first_found: clear up the skip option usage

### PR DESCRIPTION
**Pull request by ianw**
I don't think I'm the first person to be confused by the "skip"
semantics here; the option was deprecated with
e17a2b502d6601be53c60d7ba1c627df419460c9 but then restored with
707e33793d683254a511cd1ec825df86a5121feb.

This confusion seemed to stem from the two two ways this can be used;
when used with "with_first_found", setting "skip: True" lives up to
its name and makes the task skip when no files are found.

When used in a template with lookup()/query(), however, "skip: True"
returns an empty list.  You have plenty of options...

  lookup('first_found', dict(files=['notfound']), errors='ignore') -> ""
  lookup('first_found', dict(files=['notfound'], skip=True))       -> []
  lookup('first_found', dict(files=['notfound']), wantlist=True)   -> []
  q('first_found', dict(files=['notfound']), errors='ignore')      -> []

I think this is not helped by the extant error that mixes things up
saying "use errors='ignore' to skip" which, as noted, doesn't actually
skip.

This clarifies the documentation, suggesting the global
"errors='ignore'" is a better choice when using lookup/query.  It dds
more examples and reduces the error to just "no file was found" to
avoid giving wrong suggestions.

##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME
first_found

### CODE REVIEW

File: lib/ansible/plugins/lookup/first_found.py [hunk 1]
```diff
@@ -32,27 +32,54 @@
       skip:
         type: boolean
         default: False
-        description: Return an empty list if no file is found, instead of an error.
-"""
+        description:
+          When set with ``with_first_found``, causes the calling task to be skipped when no files are matched.  When set and used as
+          lookup in a template, will return an empty list if no matches are found.  Note this usage will *not* cause the task to skip.  For
+          this reason, it is reccommended to use ``errors='ignore'`` when used in a template with ``lookup`` or ``query``."""
```
Comment (Akasurde) [inline lib/ansible/plugins/lookup/first_found.py]:
```suggestion
          this reason, it is recommended to use ``errors='ignore'`` when used in a template with ``lookup`` or ``query``.
   """
```

File: lib/ansible/plugins/lookup/first_found.py [hunk 2]
```diff
@@ -32,27 +32,54 @@
...
     params:
       files:
-        - path/tasks.yaml
-        - path/other_tasks.yaml
+        - /path/to/foo.txt
+        - /path/to/bar.txt
+      # NOTE: not reccommended.  Use either
+      #  lookup('first_found', params, errors='ignore') (empty string)
+      #  lookup('first_found', params, errors='ignore', wantlist=True) (empty list)
+      #  query('first_found', params, errors='ignore')  (empty list)
```
Comment (samdoran) [inline lib/ansible/plugins/lookup/first_found.py]: Omit these lines please.

File: lib/ansible/plugins/lookup/first_found.py [hunk 3]
```diff
@@ -32,27 +32,54 @@
...
       - /path/to/foo.txt
       - bar.txt  # will be looked in files/ dir relative to role and/or play
       - /path/to/biz.txt
 
-- name: include tasks only if files exist.
-  include_tasks:
-    file: "{{ query('first_found', params) }}"
+- name: Set _found_file to the first existing file, or an empty string if no files found
+  set_fact:
+    _found_file: "{{ lookup('first_found', params, errors='ignore') }}"
```
Comment (samdoran) [inline lib/ansible/plugins/lookup/first_found.py:52]:
```suggestion
    _found_file: "{{ lookup('first_found', **params, errors='ignore') }}"
```

File: lib/ansible/plugins/lookup/first_found.py [hunk 4]
```diff
@@ -32,27 +32,40 @@
       skip:
         type: boolean
         default: False
-        description: Return an empty list if no file is found, instead of an error.
-"""
+        description:
+          When set with ``with_first_found``, causes the calling task to be skipped when no files are matched.  When set and used as
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/first_found.py]: with_ anything that returns an empty list has this effect, this is not just first_found

Comment (ianw) [reply lib/ansible/plugins/lookup/first_found.py]: OK ... do you want me to reword this or what?

Comment (bcoca) [reply lib/ansible/plugins/lookup/first_found.py]: you can just remove the first phrase, 'skip' just causes the lookup to return empty list if no file is matched

Comment (bcoca) [reply lib/ansible/plugins/lookup/first_found.py]: also this is different than errors ignore, while skip=false will produce an error (after all terms are processed) errors=ignore deals with more global issues with the lookup, not just 'non of the terms matched a file'.

Comment (ianw) [reply lib/ansible/plugins/lookup/first_found.py]: alright, i think it is still worth calling out that "skip" in template context doesn't skip.  I'll reword things and the examples.

File: lib/ansible/plugins/lookup/first_found.py [hunk 5]
```diff
@@ -32,27 +32,55 @@
...
       - /path/to/foo.txt
       - bar.txt  # will be looked in files/ dir relative to role and/or play
       - /path/to/biz.txt
 
-- name: include tasks only if files exist.
-  include_tasks:
-    file: "{{ query('first_found', params) }}"
+- name: Set _found_file to the first existing file, or an empty list if no files found
+  set_fact:
+    _found_file: "{{ lookup('first_found', params) }}"
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/first_found.py]: we are trying to move away from this syntax, use this instead:
```
_found_file: "{{ lookup('first_found', findme, paths=['/extra/path/'], skip=True) }}"
```

File: lib/ansible/plugins/lookup/first_found.py [hunk 6]
```diff
@@ -32,27 +32,55 @@
...
+    files:
+     - path/tasks.yaml
+     - path/other_tasks.yaml
+    skip: True
+
+- name: Include tasks only if one of the files exists, otherwise skip
+  include_tasks: '{{ tasks_file }}'
+  when: tasks_file != ""
+  vars:
+    tasks_file: "{{ lookup('first_found', dict(files=['tasks.yaml', 'other_tasks.yaml']), errors='ignore') }}"
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/first_found.py]: again, this syntax we should phase out
```
asks_file: "{{ lookup('first_found', files=['tasks.yaml', 'other_tasks.yaml'], errors='ignore') }}"
```

Comment (ianw) [reply lib/ansible/plugins/lookup/first_found.py]: I've done this, but does this mean that "errors", "wantlist", "allow_unsafe", etc. are now reserved words; i.e. a plugin can't also use them to mean something different?  that didn't seem to be a problem when the arguments were passed as a separate dict.

Comment (bcoca) [reply lib/ansible/plugins/lookup/first_found.py]: they always were, most lookups don't take a dict as argument, older lookups, like this one, had created their own interfaces, most others use a 'single string' k=v space separated options (also something we are moving away from)

**Review by bcoca — APPROVED** [review]

**Comment by Akasurde** [comment]
@ianw Thanks for the contribution. @bcoca @samdoran Thanks for the review.

**Pull request merged**